### PR TITLE
List support: pop method

### LINKED ISSUE #1690: pop: removes and returns the last value from the list or the given index value.
### Feature Description
Add support for pop() method for lists in the Python translation.

### Test Code
This code should work when this issue is fixed:
```python
>> a = [1, 3]
>> b = a.pop(3)
>> print(b)
>> 3
```

### Proposed Solution
See description in https://github.com/pyccel/pyccel/issues/1529

### PR DESCRIPTION
**Pull request by mustapha-belbiad**
Add support for the pop() method of Python lists to the semantic stage. Add handling of that method to the Python printer. This fixes #1690.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
Don't forget to update the AUTHORS file

**Comment by mustapha-belbiad** [comment]
> Don't forget to update the AUTHORS file
alright

**Review by Farouk-Echaref — APPROVED** [review]
All looks good

File: .gitignore [hunk 1]
```diff
@@ -68,3 +68,5 @@ tests/examples/openmp/*.mod
 tests/examples/openmp/*.o
 
 benchmarks/tmp/*
+
+my_main.py
```
Comment (EmilyBourne) [inline .gitignore:72]: I don't think it is a good idea to add files to this folder following the convention that each individual developer uses for their test files. I recommend running tests in a separate folder if you are worried about accidentally committing your tests.

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ All notable changes to this project will be documented in this file.
 -   #1685 : Add support for `type()` function with class instance argument.
 -   #1605 : Add support for class methods and interfaces (including `__init__` and `__del__`).
 -   #1680 : Add support for `typing.Final`.
+-   #1690 : Add support for pop() method for
```
Comment (EmilyBourne) [inline CHANGELOG.md:22]: This sentence is not complete.

File: pyccel/ast/basic.py [hunk 3]
```diff
@@ -66,7 +66,6 @@ def __init__(self):
         self._recursion_in_progress = False
         for c_name in self._my_attribute_nodes: #pylint: disable=not-an-iterable
             c = getattr(self, c_name)
-
```
Comment (EmilyBourne) [inline pyccel/ast/basic.py:69]: This change does not seem relevant to this PR. Please try to remove irrelevant changes from PRs so that they are as small as possible. This makes them easier to review (this will be especially important in the future when your PRs are larger)

File: pyccel/ast/builtin_objects/list_functions.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+The List container has a number of built-in methods that are 
+always available.
+
+In this module we implement List methods.
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:10]: I don't think the implementation of list methods is found in this file. Rather it contains objects which describe these methods within Pyccel's AST

File: pyccel/ast/builtin_objects/list_functions.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+"""
+
+from pyccel.ast.internals import PyccelInternalFunction
+from pyccel.ast.datatypes import NativeHomogeneousList
+
+class ListPop(PyccelInternalFunction) :
+    """
+    Represents a call to the .pop() method.
+    
+    Represents a call to the .pop() method wich
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:20]:
```suggestion
    Represents a call to the .pop() method which 
```

File: pyccel/ast/builtin_objects/list_functions.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    removes the item at the specified index. 
+    The method also returns the removed item.
+    
+    Parameters
+    ----------
+    *args : iterable
+        The arguments passed to the function call.
+    """
+    __slots__ = ('_dtype',)
+    _precision = -1
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:30]: How do you know the precision of the result?

File: pyccel/ast/builtin_objects/list_functions.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    _rank  = 0
+    _shape = None
+    _order = None
+    _class_type = NativeHomogeneousList()
+    name = 'pop'
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._dtype = args[0].dtype
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:38]: As described in the second paragraph [here](https://github.com/pyccel/pyccel/blob/devel/developer_docs/ast_nodes.md#pyccel-internal-function), objects inheriting from PyccelInternalFunction must provide a constructor whose positional and keyword arguments match the names of the positional and keyword arguments of the Python function being described.

According to [Python's docs](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists) list.append only accepts 1 argument whose name may be `i`. Your function takes any number of arguments. What are these additional arguments supposed to represent? Further the argument to `pop` is optional however this case does not seem to be handled by your implementation. This isn't a problem in the Python implementation but it will be much simpler for the C/Fortran implementations if it is already defined here.

File: pyccel/ast/core.py [hunk 8]
```diff
@@ -3383,9 +3381,7 @@ def __init__(
 
          #  methods = list(methods) + [free]
          # TODO move this somewhere else
-
         methods = tuple(methods)
-
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:3388]: Lots of unnecessary changes in this file

File: pyccel/codegen/pipeline.py [hunk 9]
```diff
@@ -169,7 +169,6 @@ def handle_error(stage):
 
     if conda_warnings not in ('off', 'basic', 'verbose'):
         raise ValueError("conda warnings accept {off, basic,verbose}")
-
```
Comment (EmilyBourne) [inline pyccel/codegen/pipeline.py:172]: Unnecessary change

File: pyccel/codegen/printing/ccode.py [hunk 10]
```diff
@@ -514,7 +514,6 @@ def _init_stack_array(self, expr):
         tot_shape = self._print(functools.reduce(
             lambda x,y: PyccelMul(x,y,simplify=True), var.alloc_shape))
         declare_dtype = self.find_in_dtype_registry(NativeInteger(), 8)
-
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:517]: Unnecessary change

File: pyccel/codegen/printing/pycode.py [hunk 11]
```diff
@@ -249,7 +249,6 @@ def _print_DottedVariable(self, expr):
     def _print_FunctionDefArgument(self, expr):
         name = self._print(expr.name)
         default = ''
-
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/pycode.py:252]: Unnecessary change. Why do you hate vertical whitespace ^^

File: pyccel/codegen/printing/pycode.py [hunk 12]
```diff
@@ -1136,6 +1135,12 @@ def _print_TypingFinal(self, expr):
         annotation = self._print(expr.arg)
         return f'const {annotation}'
 
+    def _print_ListPop(self, expr):
+        args = ','.join(self._print(a) for a in expr.args[1:])
+        name = self._print(expr.args[0])
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/pycode.py:1140]: It would be nice to give expr.args[0] a more explicit names. This can be done using the @property decorator with class methods. ` expr.args[1:]` could stay as is if you can tell me what all these arguments are for

File: pyccel/codegen/printing/pycode.py [hunk 13]
```diff
@@ -1136,6 +1135,12 @@ def _print_TypingFinal(self, expr):
...
+        method = self._print(expr.name)
+        return f"{name}.{method}({args})"
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/pycode.py:1142]: Do you foresee a case where this would be something other than:
```suggestion
        return f"{name}.pop({args})"
```
?
For example an object which inherits from `pop`? If not then I don't think that there is anything to be gained from using `self._print` which incurs a cost

File: pyccel/parser/semantic.py [hunk 14]
```diff
@@ -298,7 +298,6 @@ def annotate(self):
 
         # TODO - add settings to Errors
         #      - filename
-
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:301]: Lots of unnecessary changes in this file

File: pyccel/parser/syntactic.py [hunk 15]
```diff
@@ -323,7 +323,6 @@ def _visit(self, stmt):
         # TODO - add settings to Errors
         #      - line and column
         #      - blocking errors
-
```
Comment (EmilyBourne) [inline pyccel/parser/syntactic.py:326]: Unnecessary change

File: tests/epyccel/test_epyccel_lists.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    )
+)
+
+def test_pop_2(language) :
+    def po():
+        a = [1.7,2.7,45.0]
+        a.pop()
+        return a.pop(-1)
+    re = epyccel(po, language = language)
+    assert po() == re()
```
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_lists.py:80]: It would be good to add some tests to this file which check that the return type is as expected.

File: tests/pyccel/scripts/hope_benchmarks/test.json [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        "includes": [
+            "/usr/include/python3.11",
+            "/usr/local/lib/python3.11/dist-packages/numpy/core/include"
+        ],
+        "shared_suffix": ".cpython-311-x86_64-linux-gnu.so",
+        "dependencies": [
+            "/usr/lib/x86_64-linux-gnu/libpython3.11.so"
+        ]
+    }
+}
```
Comment (EmilyBourne) [inline tests/pyccel/scripts/hope_benchmarks/test.json:67]: This file is generated by the tests. It should not be committed to the repo

File: tests/pyccel/scripts/test2.json [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        "includes": [
+            "/usr/include/python3.11",
+            "/usr/local/lib/python3.11/dist-packages/numpy/core/include"
+        ],
+        "shared_suffix": ".cpython-311-x86_64-linux-gnu.so",
+        "dependencies": [
+            "/usr/lib/x86_64-linux-gnu/libpython3.11.so"
+        ]
+    }
+}
```
Comment (EmilyBourne) [inline tests/pyccel/scripts/test2.json:67]: This file is generated by the tests. It should not be committed to the repo

File: tests/semantic/test_semantic.py
Comment (EmilyBourne) [inline tests/semantic/test_semantic.py:1]: What did you modify in this file?

File: tests/syntax/test_syntax.py
Comment (EmilyBourne) [inline tests/syntax/test_syntax.py:1]: What did you modify in this file?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
This looks like a good start but I see a few major flaws. There are also lots and lots of whitespace changes. Everyone had different opinions on what whitespace is more readable. We don't impose any conventions in the code, you shouldn't modify lots of unrelated code to impose your preferred conventions.

File: CHANGELOG.md (hunk 2 above)
Comment (mustapha-belbiad) [reply CHANGELOG.md:22]: My apologies, I'm going to fix it

File: pyccel/codegen/printing/pycode.py (hunk 12 above)
Comment (mustapha-belbiad) [reply pyccel/codegen/printing/pycode.py:1140]: **expr.args[1:]** The purpose of this is to print multiple arguments.

Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py:1140]: Please give me an example where `pop` needs to handle multiple arguments

File: pyccel/ast/builtin_objects/list_functions.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    _order = None
+    _shape = None
+    _class_type = NativeHomogeneousList()
+    name = 'pop'
+    def __init__(self, name, index_elemnt=None):
+        if index_elemnt:
+            super().__init__(name, index_elemnt)
+        else:
+            super().__init__(name)
+        self._index = index_elemnt
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:43]: Please fix the typo:
```suggestion
    def __init__(self, name, index_elemnt=None):
        if index_elemnt:
            super().__init__(name, index_element)
        else:
            super().__init__(name)
        self._index = index_element
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -19,6 +19,7 @@ All notable changes to this project will be documented in this file.
 -   #1685 : Add support for `type()` function with class instance argument.
 -   #1605 : Add support for class methods and interfaces (including `__init__` and `__del__`).
 -   #1680 : Add support for `typing.Final`.
+-   #1690 : Add support for pop() method of list class
```
Comment (EmilyBourne) [inline CHANGELOG.md:22]: Code should be quoted
```suggestion
-   #1690 : Add support for `pop()` method of list class
```

File: pyccel/ast/builtin_objects/list_functions.py (hunk 19 above)
Comment (EmilyBourne) [reply pyccel/ast/builtin_objects/list_functions.py:44]: Not sure why this was resolved. The typo is still there

File: pyccel/ast/class_defs.py [hunk 21]
```diff
@@ -5,12 +5,14 @@
 """
 This module contains all types which define a python class which is automatically recognised by pyccel
 """
+
+from pyccel.ast.builtin_objects.list_functions import (ListPop)
```
Comment (EmilyBourne) [inline pyccel/ast/class_defs.py:9]:
```suggestion
from pyccel.ast.builtin_objects.list_functions import ListPop
```
These parentheses do nothing

File: pyccel/ast/builtin_objects/list_functions.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        self._dtype = name.dtype
+        self._precision = name.precision
+
+    @property
+    def args(self):
+        """
+        The current index value for the element to be popped.
+
+        The current index value for the element to be popped.
+        """
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:55]: Can you explain why "args" is the best name to describe "The current index value for the element to be popped.". Why not something like `pop_index`?

The only information that I can get from seeing "args" in the code is that it contains more than 1 object (as args is plural), but this isn't true :/

File: pyccel/parser/semantic.py [hunk 23]
```diff
@@ -740,6 +740,10 @@ def _infer_type(self, expr):
             d_var['memory_handling'] = 'heap'
             return d_var
 
+        elif isinstance(expr, PythonList):
+            d_var['cls_base'       ] = ListClass
+            d_var['memory_handling'] = 'heap'
+            return d_var
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:746]: Is this condition necessary? Can you explain why this is not handled by the condition below `elif isinstance(expr, TypedAstNode):`? (Maybe `PythonTuple` caused confusion? The condition `elif isinstance(expr, PythonTuple):` is removed in one of my unfinished PRs as it is no longer necessary)

File: tests/epyccel/test_epyccel_lists.py [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+def test_pop_2(language) :
+    def po():
+        a = [1.7,2.7,45.0]
+        a.pop()
+        return a.pop(-1)
+    re = epyccel(po, language = language)
+    assert isinstance(po(), type(re()))
+    assert po() == re()
```
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_lists.py:84]: This file contains lots of repeat calls. This could be problematic if the function or its wrapper is costly. Why not save the results?

Additionally the names are not very explicit. What does `re` stand for?

```suggestion
def test_pop_2(language) :
    def pop_2():
        a = [1.7,2.7,45.0]
        a.pop()
        return a.pop(-1)
    pop_2_epyc = epyccel(po, language = language)
    python_result = pop_2()
    pyccel_result = pop_2_epyc()
    assert isinstance(python_result, type(pyccel_result))
    assert python_result == pyccel_result
```

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
This looks good but I still have a few comments

File: CHANGELOG.md (hunk 20 above)
Comment (EmilyBourne) [reply CHANGELOG.md]: Fixed 540283d66e640f3f7dd9644ea25427981d31cfbe

File: pyccel/ast/class_defs.py (hunk 21 above)
Comment (EmilyBourne) [reply pyccel/ast/class_defs.py]: Fixed 8d649bfdb127979ea3223f0bfc80ba57ca5efacc

File: pyccel/ast/builtin_objects/list_functions.py (hunk 22 above)
Comment (EmilyBourne) [reply pyccel/ast/builtin_objects/list_functions.py]: Fixed 61433ec7f31d1c86e7449b3b91482fae3adfc110

File: pyccel/parser/semantic.py (hunk 23 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Fixed 7bd019398f4fa6e6b0f7050c9bbc38724fdabd2e

File: tests/epyccel/test_epyccel_lists.py (hunk 24 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_lists.py]: Fixed 8e921a7c9a042604974a2e1d53a18cfc8b3e8d3f

File: pyccel/ast/builtin_objects/list_functions.py [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+    _shape = None
+    _class_type = NativeHomogeneousList()
+    name = 'pop'
+    def __init__(self, name, index_element=None):
+        if index_element:
+            super().__init__(name, index_element)
+        else:
+            super().__init__(name)
+        self._index = index_element
+        self._list_variable = name
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:45]: You are saving `name` in an attribute called `_list_variable`. These two names seem to imply different things. Why the difference? In my opinion `list_variable` makes more sense than `name` as you are storing a `TypedAstNode` describing a list not the name of a Variable (see similar discussion here: https://github.com/pyccel/pyccel/pull/1709#discussion_r1485854955)

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
One final comment

File: pyccel/ast/builtin_objects/list_functions.py [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+    index_element : TypedAstNode
+        The current index value for the element to be popped.
+    """
+    __slots__ = ('_dtype','_precision', '_index','_list_variable')
+    _rank = 0
+    _order = None
+    _shape = None
+    _class_type = NativeHomogeneousList()
+    name = 'pop'
+    def __init__(self, last_variable, index_element=None):
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:39]: Typo `last_variable`->`list_variable`

File: pyccel/ast/builtin_objects/list_functions.py [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    _shape = None
+    _class_type = NativeHomogeneousList()
+    name = 'pop'
+    def __init__(self, list_variable, index_element=None):
+        if index_element:
+            super().__init__(list_variable, index_element)
+        else:
+            super().__init__(list_variable)
+        self._index = index_element
+        self._list_variable = list_variable
```
Comment (EmilyBourne) [inline pyccel/ast/builtin_objects/list_functions.py:45]: See the [docs](https://github.com/pyccel/pyccel/blob/devel/developer_docs/ast_nodes.md#pyccel-internal-function)

> The arguments can be passed through in the constructor of the sub-class to let PyccelInternalFunction take care of saving the arguments, or the arguments can be saved with more useful names inside the class. However care must be taken to ensure that the argument is not saved inside the class twice.

Here the argument is saved twice

File: CHANGELOG.md [hunk 28]
```diff
@@ -25,6 +25,7 @@ All notable changes to this project will be documented in this file.
 -   #1605 : Add support for methods and interfaces in classes (including `__init__` and `__del__`).
 -   #1618 : Add support for class instance attributes.
 -   #1680 : Add support for `typing.Final`.
+-   #1690 : Add support for `pop()` method of list class
```
Comment (EmilyBourne) [inline CHANGELOG.md:28]: Please move this to the new UNRELEASED section

**Comment by yguclu** [comment]
@mustapha-belbiad Please resolve the conflicts with the `devel` branch and update your branch.

File: pyccel/ast/builtin_objects/list_functions.py
Comment (yguclu) [inline pyccel/ast/builtin_objects/list_functions.py:1]: Please add the `__all__` tuple to this file.

File: pyccel/codegen/printing/pycode.py [hunk 29]
```diff
@@ -1136,6 +1136,13 @@ def _print_TypingFinal(self, expr):
         annotation = self._print(expr.arg)
         return f'const {annotation}'
 
+    def _print_ListPop(self, expr):
+        args = ""
+        if expr.pop_index:
+            args = self._print(expr.pop_index)
```
Comment (yguclu) [inline pyccel/codegen/printing/pycode.py:1142]: Small suggestion. Instead of redefining the variable `args`, it is more Pythonic to use the single-liner
```suggestion
        args = self._print(expr.pop_index) if expr.pop_index else ""
```

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job! I have a few minor comments.

Do not forget to resolve the conflicts with `devel`!

File: pyccel/ast/builtin_objects/list_functions.py
Comment (EmilyBourne) [reply pyccel/ast/builtin_objects/list_functions.py:1]: Beware, the name of this file changed in #1709 you should also move your functions when you merge devel into your branch

**Comment by EmilyBourne** [comment]
@mustapha-belbiad Don't forget to react to the comments on your PR. See the [docs](https://github.com/pyccel/pyccel/blob/devel/developer_docs/review_process.md):

> Once your pull request has been reviewed please react to the open conversations. If you disagree you can say this, if not please leave a reference to the commit which fixes the mentioned issue. This makes the review process faster and more streamlined. Please only resolve conversations that you opened. You may think you fixed the problem, but the reviewer may disagree and leaving the discussion open makes it easier for them to verify that they agree with you. If you are reviewing then please close all conversations that you open once the problem is resolved. If you don't this can block the merge.

File: pyccel/codegen/printing/pycode.py (hunk 12 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/pycode.py]: Pyccel only works properly if your code runs properly in Python. `.pop(1,3,4)` will raise an error when you run your code in Python so we can safely ignore this case in Pyccel.

**Review by yguclu — APPROVED** [review]
